js/apple2.ts

```diff
@@ -76,40 +80,56 @@ export class Apple2 implements Restorable<State>, DebuggerContainer {
         renderedFrames: 0
     };
 
+    public ready: Promise<[{ default: ROMConstructor }, { default: rom }, void]>;
+
     constructor(options: Apple2Options) {
         const LoresPage = options.gl ? LoresPageGL : LoresPage2D;
         const HiresPage = options.gl ? HiresPageGL : HiresPage2D;
         const VideoModes = options.gl ? VideoModesGL : VideoModes2D;
 
+        const romImportPromise = import(`./roms/system/${options.rom}`);
```

Comment:
Personally, I'd stick this in an `async` method like:
```typescript
constructor(options: Apple2Options) {
    ...
    this.ready = this.init(options);
}

async init(options: Apple2Options) {
    const romImportPromise = import(`./roms/system/${options.rom}`);
    const characterRomImportPromise = import(`./roms/character/${options.characterRom}`);
    const [{ default: Apple2ROM }, { default: characterRom }] = await Promise.all([
        romImportPromise,
        characterRomImportPromise,
        this.vm.ready,
    ]);

    this.rom = new Apple2ROM();
    ...
}
```

The main reason, in my mind, is that async functions are a bit easier to think about than Promise chains.

The second reason is that I think this `Promise` implementation is broken because `this.ready` is the result of the `Promise.all` call and not the result of `this.ready.then(...)` setup. Thus if something outside did `apple2.ready.then(...)`, the two would race and it's possible that none of the RAM/ROM/MMU would be set up yet.

Reply:
Good point. This also reminded me that these are all hanging promises that I meant to triage.